### ISSUE TITLE
Storing dict keys during traversal

### ISSUE THREAD

**Issue by therve** [mahmoud/glom#276]
Hi,

I'm struggling to do what seems to be a simple thing when reformatting data, basically store the grandparent dict key when iterating. A simple example:

```python
from glom import *
data = {"a": {"b": {"c": 1}}}

print(
    glom(data, (
          T.values(),
          Iter(T.values()),
          Flatten(),
          [{"value": "c"}]
      )                                                                                                                
      )
  )  
```

gives me `[{'value': 1}]`  as I expect. What I'd like is to have `[{'value': 1, 'path': 'a'}]`. Any idea how to to achieve this? I tried to use `S` but I never seems to get the scope right. It seems I'm missing something obvious. Thanks!

**Comment by mahmoud** [comment]
Ah, the _grandparent_ key, sure! This isn't as pretty as I would like, but here's the working spec:

```python
(
        T.items(), 
        Iter(
          (A.globals.gp_item, T[1].values(), 
           Iter({'val': 'c', 
                 'path': S.globals.gp_item[0]}))
        ).flatten().all()
)
```

And I get:
```
[
    {
        "val": 1,
        "path": "a"
    }
]
```


[Here's a link to a glompad where you can fool around with it in your browser](https://yak.party/glompad/#spec=%28%0A++++++++T.items%28%29%2C+%0A++++++++Iter%28%0A++++++++++%28%28A.globals.gp_item%2C+T%5B1%5D%29%2C+T.values%28%29%2C+%0A+++++++++++Iter%28%7B%27val%27%3A+%27c%27%2C+%0A+++++++++++++++++%27path%27%3A+S.globals.gp_item%5B0%5D%7D%29%29%0A++++++++%29.flatten%28%29.all%28%29%0A%29&target=%7B%22a%22%3A+%7B%22b%22%3A+%7B%22c%22%3A+1%7D%7D%7D&v=1).

Two key things to note:

* I'm using [the `globals` member](https://glom.readthedocs.io/en/latest/api.html#sensible-saving-vars-s-globals) of the scope (and [the `A` assignment shortcut](https://glom.readthedocs.io/en/latest/api.html#updating-the-scope-s-a)) to save the grandparent item before we descend. Probably when you used `S` the value was being lost because the scope was local.
    * If this is part of a much larger spec, you can create your own non-local, non-global scope with `Vars` documented in the link above.
    * I found it simpler to save the grandparent item (key and value), instead of just the value, and then only emit the key (`gp_item[0]`).
*  Notice I'm using two `Iter` specs. glom doesn't yet have full tree traversal, so the nesting of the spec is going to roughly match the nesting of your data. If you need this to work in an arbitrary nested structure, I recommend [`remap`](https://sedimental.org/remap.html), the recursive map function, part of [`boltons.iterutils`](https://boltons.readthedocs.io/en/latest/iterutils.html#nested). 

Hope this helps!

**Comment by therve** [comment]
Thanks a lot, I almost reached that one myself! For some reason I was focused on trying to save just the key `T[0]` instead of the whole `items` structure, this works.

And a related thank you for that great library.

**Comment by mahmoud** [comment]
Yeah, I know what you mean re: keys vs items. I might add this example to the docs, in case it's instructive for others too :)
